pandas/core/series.py

```diff
@@ -4100,8 +4107,8 @@ def _try_cast(arr, take_fast_path):
             elif is_extension_array_dtype(dtype):
                 # create an extension array from its dtype
                 array_type = dtype.construct_array_type()._from_sequence
-                subarr = array_type(subarr, dtype=dtype, copy=copy)
```

Comment:
what is the issue here?

Reply:
For some reason, I get a `NameError` 

```python
In [2]: pd.Series([0, 1, 2], dtype=pd.core.sparse.dtype.SparseDtype(int, 0))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/sandbox/pandas/pandas/core/series.py in _try_cast(arr, take_fast_path)
   4088             if is_float_dtype(dtype) or is_integer_dtype(dtype):
-> 4089                 subarr = maybe_cast_to_integer_array(arr, dtype)
   4090

~/sandbox/pandas/pandas/core/dtypes/cast.py in maybe_cast_to_integer_array(arr, dtype, copy)
   1342         if not hasattr(arr, "astype"):
-> 1343             casted = np.array(arr, dtype=dtype, copy=copy)
   1344         else:

TypeError: data type not understood

During handling of the above exception, another exception occurred:

UnboundLocalError                         Traceback (most recent call last)
<ipython-input-2-a1217d9421a3> in <module>()
----> 1 pd.Series([0, 1, 2], dtype=pd.core.sparse.dtype.SparseDtype(int, 0))

~/sandbox/pandas/pandas/core/series.py in __init__(self, data, index, dtype, name, copy, fastpath)
    280             else:
    281                 data = _sanitize_array(data, index, dtype, copy,
--> 282                                        raise_cast_failure=True)
    283
    284                 data = SingleBlockManager(data, index, fastpath=True)

~/sandbox/pandas/pandas/core/series.py in _sanitize_array(data, index, dtype, copy, raise_cast_failure)
   4154         if dtype is not None:
   4155             try:
-> 4156                 subarr = _try_cast(data, False)
   4157             except Exception:
   4158                 if raise_cast_failure:  # pragma: no cover

~/sandbox/pandas/pandas/core/series.py in _try_cast(arr, take_fast_path)
   4109                 array_type = dtype.construct_array_type()._from_sequence
   4110                 # XXX: this needs re-working.
-> 4111                 subarr = array_type(subarr, dtype=dtype, copy=copy)
   4112             elif dtype is not None and raise_cast_failure:
   4113                 raise

UnboundLocalError: local variable 'subarr' referenced before assignment
```

not sure why our other tests don't catch this, but I *think* we want `arr` here anyway.

Reply:
hmm odd. i guess we don't hit this, though i thought we did. any repro outside of sparse?

Reply:
The relevant conditions seem to be that

- `dtype` must be an `ExtensionDtype`
- `dtype` must be considered float or integer by `is_float_dtype` / `is_integer_dtype`

which would explain why we haven't hit it before, as we don't have any EAs that should be considered float or integer.

If those are true, then we call `maybe_cast_to_integer_array`, which fails for non-numpy dtypes. Maybe we should just skip that call to `maybe_cast_to_integer_array` when we have an extension dtype, since we won't want to do that casting anyway.